### PR TITLE
AxisTickCalculator_Date: Make sure we exit if we tried all possible timespans

### LINKED ISSUE #601: IndexOutOfBoundsException introduced in 3.7.0 or 3.8.0
Hi, attached date+value series can be plotted in 3.6.6 without any errors. Yet in 3.8.0 (cannot use 3.7.0 as this release doesn't contain some of the class that 3.6.6 and 3.8.0 have) I got:

`java.lang.IndexOutOfBoundsException: Index 46 out of bounds for length 46
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
	at java.base/java.util.Objects.checkIndex(Objects.java:372)
	at java.base/java.util.ArrayList.get(ArrayList.java:458)
	at org.knowm.xchart.internal.chartpart.AxisTickCalculator_Date.calculate(AxisTickCalculator_Date.java:201)
	at org.knowm.xchart.internal.chartpart.AxisTickCalculator_Date.<init>(AxisTickCalculator_Date.java:103)
	at org.knowm.xchart.internal.chartpart.Axis.getAxisTickCalculator(Axis.java:454)
	at org.knowm.xchart.internal.chartpart.Axis.getXAxisHeightHint(Axis.java:289)
	at org.knowm.xchart.internal.chartpart.Axis.preparePaint(Axis.java:155)
	at org.knowm.xchart.internal.chartpart.AxisPair.paint(AxisPair.java:121)
	at org.knowm.xchart.XYChart.paint(XYChart.java:416)
	at org.knowm.xchart.BitmapEncoder.getBufferedImage(BitmapEncoder.java:281)
	at org.knowm.xchart.BitmapEncoder.saveBitmap(BitmapEncoder.java:134)
	at org.knowm.xchart.BitmapEncoder.saveBitmap(BitmapEncoder.java:109)
`

which corresponds to unverified use of index in a loop here:

`double gridStep =
          timeSpans.get(++index).getUnitAmount()
              * timeSpans.get(index).getMagnitude(); // in time units (ms)
              *
`

This error most probably has been introduced as part of this fix: https://github.com/knowm/XChart/issues/527 (@timmolter)

example data that leads to this error:

[in.csv](https://github.com/knowm/XChart/files/6569374/in.csv)

### PR DESCRIPTION
**Pull request by Flole998**
Unfortunately I don't know how to add a proper test for this but in #601 there is test data that can be used for that.

Fixes #601

### CODE REVIEW

**Comment by J-N-K** [comment]
@timmolter Any chance to get this merged and released? Would be great to be able to use the new version but unfortunately this is a blocker for us.

**Comment by Flole998** [comment]
@mccartney Since you apparently recently merged some PRs, could you please have a look at this one aswell and merge it?

File: xchart/src/main/java/org/knowm/xchart/internal/chartpart/AxisTickCalculator_Date.java [hunk 1]
```diff
@@ -242,6 +242,9 @@ protected void calculate() {
         // }
       }
       //      System.out.println("************");
+      if(index >= timeSpans.size() - 1)
+        break; // We reached the end, even though we might not yet have the result we want,
+               // continuing will lead to an Exception!
     } while (skip
         || !areAllTickLabelsUnique(tickLabels)
         || !willLabelsFitInTickSpaceHint(tickLabels, gridStepInChartSpace));
```
Comment (mccartney) [inline xchart/src/main/java/org/knowm/xchart/internal/chartpart/AxisTickCalculator_Date.java:250]: A tiny remark is that a break at the end of the loop can likely be replaced by changing the condition in `while`. Can you try that?

Comment (Flole998) [reply xchart/src/main/java/org/knowm/xchart/internal/chartpart/AxisTickCalculator_Date.java:250]: The condition will get a lot more complicated and the explanation comment why this is necessary probably needs to be removed. Are you sure that's a good idea?

Comment (mccartney) [reply xchart/src/main/java/org/knowm/xchart/internal/chartpart/AxisTickCalculator_Date.java:250]: I am not sure, but in general the readability of that loop is not perfect.

Comment (mccartney) [reply xchart/src/main/java/org/knowm/xchart/internal/chartpart/AxisTickCalculator_Date.java:250]: Maybe we can just merge as is, and then clean-up the code afterwards.

Comment (J-N-K) [reply xchart/src/main/java/org/knowm/xchart/internal/chartpart/AxisTickCalculator_Date.java:250]: Do you have an idea when there will be a new release?

Comment (mccartney) [reply xchart/src/main/java/org/knowm/xchart/internal/chartpart/AxisTickCalculator_Date.java:250]: I don't. AFAIK @timmolter is the only one who can issue a new release.

**Comment by timmolter** [comment]
@mccartney Would now be a good time for a release in your opinion?

**Comment by mccartney** [comment]
Yes. If possible, please do the needful.

**Pull request merged**